subt/main.py

```diff
@@ -541,26 +597,94 @@ def play_system_track(self):
-        self.collision_detector_enabled = True
-        self.follow_wall(radius = 0.9, right_wall=self.use_right_wall,
-                            timeout=timedelta(minutes=12, seconds=0))
-        self.collision_detector_enabled = False
+    def play_virtual_part(self):
+        self.stdout("Waiting for origin ...")
+        self.origin = None  # invalidate origin
+        self.origin_error = False
+        self.bus.publish('request_origin', True)
+        while self.origin is None and not self.origin_error:
+            self.update()
+        self.stdout('Origin:', self.origin, self.robot_name)
+
+        if self.origin is not None:
+            x, y, z = self.origin
+            x1, y1, z1 = self.xyz
+            self.offset = x - x1, y - y1, z - z1
+            self.stdout('Offset:', self.offset)
+            heading = quaternion.heading(self.origin_quat)
+            self.stdout('heading', math.degrees(heading), 'angle', math.degrees(math.atan2(-y, -x)), 'dist', math.hypot(x, y))
+
+            self.turn(normalizeAnglePIPI(math.atan2(-y, -x) - heading))
+            self.go_straight(math.hypot(x, y))  # go to the tunnel entrance
+        else:
+            # lost in tunnel
+            self.stdout('Lost in tunnel:', self.origin_error, self.offset)
+        for loop in range(3):
+            self.collision_detector_enabled = True
+            dist, reason = self.follow_wall(radius=self.walldist, right_wall=self.use_right_wall,  # was radius=0.9
+                                timeout=self.timeout, pitch_limit=LIMIT_PITCH, roll_limit=None)
+            self.collision_detector_enabled = False
+            if reason is None or reason != REASON_PITCH_LIMIT:
+                break
+            self.stdout(self.time, "Microstep HOME %d %.3f" % (loop, dist), reason)
+            self.go_straight(-0.3, timeout=timedelta(seconds=10))
+            self.return_home(timedelta(seconds=10))
 
-        print("Artifacts:", self.artifacts)
+        self.stdout("Artifacts:", self.artifacts)
 
-        print("Going HOME")
-        self.return_home()
+        self.stdout(self.time, "Going HOME %.3f" % dist, reason)
 
+        self.return_home(2 * self.timeout)
         self.send_speed_cmd(0, 0)
 
         if self.artifacts:
             self.bus.publish('artf_xyz', [[artifact_data, round(x*1000), round(y*1000), round(z*1000)] 
                                           for artifact_data, (x, y, z) in self.artifacts])
 
-        self.wait(timedelta(seconds=30))
+        self.wait(timedelta(seconds=10), use_sim_time=True)
+
+    def dumplog(self):
+        import os
+        filename = self.bus.logger.filename  # deep hack
+        self.stdout("Dump Log:", filename)
+        size = statinfo = os.stat(filename).st_size
+        self.stdout("Size:", size)
+        with open(filename, 'rb') as f:
+            for i in range(0, size, 100):
+                self.stdout(i, f.read(100))
+        self.stdout("Dump END")
```

Comment:
Do we still need this?

Reply:
I would remove it as soon as I confirm that ROS bag with `/robot_data` works and we can build on it.